pkg/apis/serving/v1alpha1/service_types.go

```diff
@@ -0,0 +1,156 @@
+/*
+Copyright 2019 Kubeflow Community.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ServiceSpec defines the desired state of Service
+type ServiceSpec struct {
+	// If specified, passes name to generated Revision.
+	RevisionName string `json:"rollout,omitempty"`
+	// Must contain either one or two elements.
+	// The first is the "active" revisions and the second is the "candidate" revision.
+	Revisions []string `json:"revisions,omitempty"`
```

Comment:
Should we use a struct instead of a slice? For example:

```go
type Revisions struct {
    Active string
    Candidate string
}
```

I think we should not use comments to define rules.

Reply:
I'm not actually super familiar with how that would work. What happens if the slice only has one element? If it makes that element "active" and keeps candidate "null" and the slice's length is 1, then this is awesome and I'll absolutely merge in.

I think the types file will be a useful place for users to dig into the exact semantic. This is fundamental and baked into the spec, not an implementation detail.